consensus/reactor_test.go

```diff
@@ -3,623 +3,822 @@ package consensus
-	waitForAndValidateBlock(t, nVals, activeVals, blocksSubs, css)
-
-	if css[0].GetRoundState().LastValidators.TotalVotingPower() == previousTotalVotingPower {
-		t.Fatalf(
-			"expected voting power to change (before: %d, after: %d)",
-			previousTotalVotingPower,
-			css[0].GetRoundState().LastValidators.TotalVotingPower())
-	}
+	rts.sub = blocksSub
 
-	updateValidatorTx = kvstore.MakeValSetChangeTx(val1PubKeyABCI, 2)
-	previousTotalVotingPower = css[0].GetRoundState().LastValidators.TotalVotingPower()
-
-	waitForAndValidateBlock(t, nVals, activeVals, blocksSubs, css, updateValidatorTx)
-	waitForAndValidateBlockWithTx(t, nVals, activeVals, blocksSubs, css, updateValidatorTx)
-	waitForAndValidateBlock(t, nVals, activeVals, blocksSubs, css)
-	waitForAndValidateBlock(t, nVals, activeVals, blocksSubs, css)
-
-	if css[0].GetRoundState().LastValidators.TotalVotingPower() == previousTotalVotingPower {
-		t.Fatalf(
-			"expected voting power to change (before: %d, after: %d)",
-			previousTotalVotingPower,
-			css[0].GetRoundState().LastValidators.TotalVotingPower())
+	// simulate handle initChain in handshake
+	if cs.state.LastBlockHeight == 0 {
+		require.NoError(t, cs.blockExec.Store().Save(cs.state))
 	}
 
-	updateValidatorTx = kvstore.MakeValSetChangeTx(val1PubKeyABCI, 26)
-	previousTotalVotingPower = css[0].GetRoundState().LastValidators.TotalVotingPower()
+	require.NoError(t, rts.reactor.Start())
+	require.True(t, rts.reactor.IsRunning())
 
-	waitForAndValidateBlock(t, nVals, activeVals, blocksSubs, css, updateValidatorTx)
-	waitForAndValidateBlockWithTx(t, nVals, activeVals, blocksSubs, css, updateValidatorTx)
-	waitForAndValidateBlock(t, nVals, activeVals, blocksSubs, css)
-	waitForAndValidateBlock(t, nVals, activeVals, blocksSubs, css)
+	t.Cleanup(func() {
+		require.NoError(t, cs.eventBus.Stop())
+		require.NoError(t, rts.reactor.Stop())
+		require.False(t, rts.reactor.IsRunning())
+	})
 
-	if css[0].GetRoundState().LastValidators.TotalVotingPower() == previousTotalVotingPower {
-		t.Fatalf(
-			"expected voting power to change (before: %d, after: %d)",
-			previousTotalVotingPower,
-			css[0].GetRoundState().LastValidators.TotalVotingPower())
-	}
+	return rts
 }
 
-func TestReactorValidatorSetChanges(t *testing.T) {
-	nPeers := 7
-	nVals := 4
-	css, _, _, cleanup := randConsensusNetWithPeers(
-		nVals,
-		nPeers,
-		"consensus_val_set_changes_test",
-		newMockTickerFunc(true),
-		newPersistentKVStoreWithPath)
-
-	defer cleanup()
-	logger := log.TestingLogger()
-
-	reactors, blocksSubs, eventBuses := startConsensusNet(t, css, nPeers)
-	defer stopConsensusNet(logger, reactors, eventBuses)
-
-	// map of active validators
-	activeVals := make(map[string]struct{})
-	for i := 0; i < nVals; i++ {
-		pubKey, err := css[i].privValidator.GetPubKey()
-		require.NoError(t, err)
-		activeVals[string(pubKey.Address())] = struct{}{}
+func simulateRouter(primary *reactorTestSuite, suites []*reactorTestSuite, dropChErr bool) {
```

Comment:
I think this should use an actual router with in-memory transports instead, since these are basically integration tests. I'll be writing some more test infrastructure for setting up these kinds of networks shortly.

Reply:
Perhaps. I have all reactors using this kind of approach. Once we have an in memory router setup that I can leverage, I'd be happy to switch it over to that 👍

Reply:
As discussed elsewhere, I think we should primarily rely on unit tests rather than integration tests for reactors, since they're much more specific, can cover more ground with less effort, and are less prone to race conditions. But we'll need some of both, obviously.

Reply:
Do we have an in-mem transport/router that I can already leverage? If so, can you point me to its use?

Reply:
There's some stuff in `p2p/router_test.go`, but we need some better convenience test infra for it, and I'm currently overhauling the transports a bit anyway. Will have them ready over the weekend.

Reply:
I've added some test tooling in `p2p/p2ptest`, including setting up in-memory networks for testing and a bunch of channel-based test assertions. This has mostly been geared towards router testing, and needs to be extended and/or adapted to reactor-based testing, but it's a start. You can see them used e.g. in `p2p/router_test.go`.